pkg/notification/notice.go

```diff
@@ -0,0 +1,170 @@
+package notification
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/aquasecurity/trivy/pkg/log"
+	"github.com/aquasecurity/trivy/pkg/version/app"
+)
+
+type versionInfo struct {
+	LatestVersion string    `json:"latest_version"`
+	LatestDate    time.Time `json:"latest_date"`
+}
+
+type announcement struct {
+	FromDate     time.Time `json:"from_date"`
+	ToDate       time.Time `json:"to_date"`
+	Announcement string    `json:"announcement"`
+}
+
+type updateResponse struct {
+	Trivy         versionInfo    `json:"trivy"`
+	Announcements []announcement `json:"announcements"`
+	Warnings      []string       `json:"warnings"`
+}
+
+type VersionChecker struct {
+	updatesApi      string
+	skipUpdateCheck bool
+	quiet           bool
+	disableMetrics  bool
+
+	done             bool
+	responseReceived bool
+	currentVersion   string
+	latestVersion    updateResponse
+}
+
+// NewVersionChecker creates a new VersionChecker with the default
+// updates API URL. The URL can be overridden by passing an Option
+// to the NewVersionChecker function.
+func NewVersionChecker(opts ...Option) *VersionChecker {
+	v := &VersionChecker{
+		updatesApi:     "https://api.trivy.cloud/updates",
+		currentVersion: app.Version(),
+	}
+
+	for _, opt := range opts {
+		opt(v)
+	}
+	return v
+}
+
+// RunUpdateCheck makes a best efforts request to determine the
+// latest version and any announcements
+// Logic:
+// 1. if skipUpdateCheck is true AND metricsDisabled are both true, skip the request
+// 2. if skipUpdateCheck is true AND metricsDisabled is false, run check with metric details but suppress output
+// 3. if skipUpdateCheck is false AND metricsDisabled is true, run update check but don't send any metric identifiers
+func (v *VersionChecker) RunUpdateCheck(ctx context.Context, args []string) {
+	logger := log.WithPrefix("notification")
+
+	if v.skipUpdateCheck && v.disableMetrics {
+		logger.Debug("Skipping update check and metric ping")
+		return
+	}
+
+	go func() {
+		args = getFlags(args)
+		client := &http.Client{
+			Timeout: 3 * time.Second,
+		}
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, v.updatesApi, http.NoBody)
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Failed to create a request for Trivy api: %v", err))
+			return
+		}
+
+		// if the user hasn't disabled metrics, send the anonymous information as headers
+		if !v.disableMetrics {
+			req.Header.Set("-x-trivy-identifier", uniqueIdentifier())
+			req.Header.Set("-x-trivy-command", strings.Join(args, " "))
+			req.Header.Set("-x-trivy-os", runtime.GOOS)
+			req.Header.Set("-x-trivy-arch", runtime.GOARCH)
+		}
+
+		req.Header.Set("User-Agent", fmt.Sprintf("trivy/%s", v.currentVersion))
+		resp, err := client.Do(req)
+		if err != nil || resp.StatusCode != http.StatusOK {
+			logger.Debug(fmt.Sprintf("Failed getting response from Trivy api: %v", err))
```

Comment:
We have a dedicated function to add error attributes.

```suggestion
			logger.Debug("Failed getting response from Trivy api", log.Err(err))
```